### PR TITLE
`Text Exercises`: Replace feedback modal with inline feedback view

### PR DESCRIPTION
**Pull request by EneaGore**
> Deploy only to TS1

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
~- [ ] I **strictly** followed the principle of **data economy** for all database calls.~
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
~- [ ] I added pre-authorization annotations according to the [guidelines]~(https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The previous PR in [Athena Feedback for Text Exercises](https://github.com/ls1intum/Artemis/pull/9241) added the option to enable preliminary AI Feedback for students. This PR improves the way the feedback is viewed by adding a new route. Enhanced error handling, integration tests and some minor bug fixes are also included.

### Description

Create new route to view the (preliminary) results inline on the text submission. Add the timeline to that view and make results not show the modal but use that route for text exercises.

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Text Exercise

1. Open a text exercise with athena feedback requests avilable
2. Request feedback from inside the editor and from the overview
3. Click on results and check that the correct submission is shown
4. Assessment should continue working as usual

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Performance Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage

✅ ❌
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| header-participation-page.component.ts | 100% |  |
| exercise-scores.component.ts | 95.59% |  |
| feedback.component.ts | 95% |  |
| rating.component.ts | 91.66% |  |
| result.component.ts | 88.81% |  |
| result.service.ts | 90.83% |  |
| result.utils.ts | 89.18% |  |
| text-editor.component.ts | 82.12% |  |
| text-editor.route.ts | 100% |  |
| course-exercise-details.component.ts | 88.37% |  |
| exercise-details-student-actions.component.ts | 86.18% |  |

### Screenshots

newAIButton
view
Flow

https://github.com/user-attachments/assets/8f9a89f4-9844-452a-b1a5-154deefa552a

## Release Notes

- **New Features**
  - Enhanced functionality for retrieving non-Athena results in submissions, ensuring only valid results are returned.
  - Streamlined feedback generation process with detailed tracking of text segments and support for early feedback requests.
  - Implemented a new method for students to request automatic non-graded feedback.
  - Enhanced user interface for text exercises with improved button visibility and history viewing options.
  - Updated labels and tooltips for feedback request features to clarify context.

- **Bug Fixes**
  - Improved error handling during feedback generation to provide clearer user feedback.
  - Enhanced filtering logic for displaying successful submission results based on exam mode.

- **Documentation**
  - Updated localization files to include new keys for improved user interaction.

- **Tests**
  - Expanded test coverage for feedback suggestion toggling and submission handling scenarios.
  - Added new tests for various submission conditions and feedback request validations.

### CODE REVIEW

**Review by eceeeren — CHANGES_REQUESTED** [review]
Tested on TS1, I can ask for feedback in the code editor but I don't see the option on the Overview. See picture:

Screenshot 2024-09-15 at 17 48 16

**Review by coolchock — DISMISSED** [review]
tested on ts1, LGTM

**Review by eceeeren — CHANGES_REQUESTED** [review]
I am not sure if it is only me but I can not see the AI Feedback Analysis anymore:

Screenshot 2024-09-15 at 20 19 36

**Review by BBesrour — COMMENTED** [review]
Tested on TS1, works as expected if you set a due date. If due date is not set, you get this warning which I think is misleading. Is this wanted?
image

**Review by coolchock — APPROVED** [review]
re-approve

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
In the exercise creation/edit views, the buttons to enable the feedback are only shown after the change detection is triggered. I think they should be shown from the start, not only after some change is triggered.

image

Apart from that, generating the AI feedback works perfectly :+1:

**Review by eceeeren — DISMISSED** [review]
Tested on a testing session, approved

**Review by SimonEntholzer — DISMISSED** [review]

**Review by edkaya — APPROVED** [review]
Tested on ts1, works expected

File: src/main/webapp/app/exercises/text/participate/text-editor.component.ts [hunk 1]
```diff
@@ -131,7 +178,12 @@ export class TextEditorComponent implements OnInit, OnDestroy, ComponentCanDeact
...
 
-    private updateParticipation(participation: StudentParticipation) {
+    /**
+     * Updates the participation, the submission selected can be chosen through submissionId, default undefined means latest
+     * @param participation The participation data
+     * @param submissionId The id of the submission of choice. undefined value defaults to the latest submission
+     */
+    private updateParticipation(participation: StudentParticipation, submissionId: number | undefined = undefined) {
         this.participation = participation;
         this.textExercise = this.participation.exercise as TextExercise;
```
Comment (FelixTJDietrich) [inline src/main/webapp/app/exercises/text/participate/text-editor.component.ts:188]: "Cannot read properties of undefined (reading 'exercise')"

Comment (FelixTJDietrich) [reply src/main/webapp/app/exercises/text/participate/text-editor.component.ts:188]: I think this happens after I vote "How useful is this feedback to you?" and refresh on an older result

**Review by FelixTJDietrich — CHANGES_REQUESTED** [review]

**Review by FelixTJDietrich — CHANGES_REQUESTED** [review]
There is a issue with the "How useful is this feedback to you?" prompt, it does not work for the older results when I refresh for example. I get an internal server error, you might need to provide also the submission id 🤔 

The rest of the functionality works as expected on TS1 :)

Code looks good except for some minor issues I found. You fixed a bunch of stuff on the server side but there are no new tests / test changes. Maybe we need some more server tests to cover those. 

Can you also fill out the "Test Coverage" section of the PR? There is a provided pythin script to automatically generate one.

**Review by FelixTJDietrich — COMMENTED** [review]
The rating works now :) and the rest too except for some internal server error in initial submit now, is this because of test server 1? 

image (5)
image (4)

```"detail": "could not execute batch [Cannot delete or update a parent row: a foreign key constraint fails (`artemis`.`participant_score`, CONSTRAINT `fk_participant_score_last_rated_result_id` FOREIGN KEY (`last_rated_result_id`) REFERENCES `result` (`id`) ON DELETE RESTRICT ON UPDATE RESTRICT)] [delete from result where id=?]; SQL [delete from result where id=?]; constraint [null]",
```

File: src/main/webapp/app/exercises/text/participate/text-editor.component.ts [hunk 2]
```diff
@@ -324,4 +411,66 @@ export class TextEditorComponent implements OnInit, OnDestroy, ComponentCanDeact
...
+            error: (error) => {
+                this.alertService.error(`artemisApp.${error.error.entityName}.errors.${error.error.errorKey}`);
+            },
+        });
+    }
+
+    /**
+     * Checks if the conditions for requesting automatic non-graded feedback are satisfied.
+     * The student can request automatic non-graded feedback under the following conditions:
+     * 1. They have a graded submission.
```
[the remark this answers is not in the document]
Comment (EneaGore) [reply src/main/webapp/app/exercises/text/participate/text-editor.component.ts:436]: updated it

**Review by eceeeren — APPROVED** [review]
Tested on TS1, working as expected, reapprove

**Review by BBesrour — APPROVED** [review]
Tested on TS1, lgtm. thnx for adding the change requests

**Comment by FelixTJDietrich** [comment]
Tested on TS1 and it does not work as expected 
image

**Comment by FelixTJDietrich** [comment]
This was after creating a new exercise

**Comment by FelixTJDietrich** [comment]
The error happens when the submission is empty:
image

Also a gap between the buttons would be nice when the screen size is slightly smaller:
image

**Pull request closed without merging**